pkg/agent/openflow/pipeline.go

```diff
@@ -825,32 +967,195 @@ func (c *client) l3ToExternalFlows(nodeIP net.IP, localSubnet net.IPNet, outputP
+	return c.pipeline[endpointDNATTable].BuildFlow(priorityNormal).
+		Cookie(c.cookieAllocator.Request(cookie.Service).Raw()).
+		MatchProtocol(protocol).
+		MatchReg(int(serviceIPReg), ipVal).
+		MatchRegRange(int(servicePortReg), unionVal, binding.Range{0, 31}).
+		Action().CT(true, egressRuleTable, ctZone).
+		DNAT(
+			&binding.IPRange{StartIP: endpointIP, EndIP: endpointIP},
+			&binding.PortRange{StartPort: endpointPort, EndPort: endpointPort},
+		).
+		LoadToMark(serviceCTMark).
+		CTDone().
+		Done()
+}
+
+// endpointHairpinFlow generates the flow which does DNAT for Service
+// hairpin packets and loads the hairpin mark to markReg.
+func (c *client) endpointHairpinFlow(endpointIP net.IP) binding.Flow {
+	return c.pipeline[conntrackCommitTable].BuildFlow(priorityNormal).
+		Cookie(c.cookieAllocator.Request(cookie.Service).Raw()).
+		MatchProtocol(binding.ProtocolIP).
+		MatchDstIP(endpointIP).
+		MatchSrcIP(endpointIP).
+		Action().SetSrcIP(hairpinIP).
+		Action().LoadRegRange(int(marksReg), hairpinMark, hairpinMarkRange).
+		Action().GotoTable(l2ForwardingOutTable).
+		Done()
+}
+
+// localEndpointForwardFlow generates the flow to set the MAC address of the Endpoint
+// which is located on current Node.
+func (c *client) localEndpointForwardFlow(podIP net.IP, podMAC net.HardwareAddr) binding.Flow {
+	return c.pipeline[localEndpointForwardingTable].BuildFlow(priorityNormal).MatchProtocol(binding.ProtocolIP).
+		MatchCTMark(serviceCTMark).
+		MatchDstIP(podIP).
+		Action().SetDstMAC(podMAC).
+		Action().DecTTL().
+		Action().GotoTable(l2ForwardingCalcTable).
+		Cookie(c.cookieAllocator.Request(cookie.Pod).Raw()).
+		Done()
+}
+
+// serviceEndpointGroup creates/modify the group/buckets of Endpoints. If the
+// withSessionAffinity is true, then buckets will resubmit packets back to
+// serviceLBTable to trigger the learn flow, the learn flow will then send packets
+// to endpointDNATTable. Otherwise, buckets will resubmit packets to
+// endpointDNATTable directly.
+func (c *client) serviceEndpointGroup(groupID binding.GroupIDType, withSessionAffinity bool, endpoints ...k8sproxy.Endpoint) binding.Group {
+	group := c.bridge.CreateGroup(groupID).ResetBuckets()
+	var resubmitTableID binding.TableIDType
+	var lbResultMark uint32
+	if withSessionAffinity {
+		resubmitTableID = serviceLBTable
+		lbResultMark = marksRegServiceNeedLearn
+	} else {
+		resubmitTableID = endpointDNATTable
+		lbResultMark = marksRegServiceHitted
+	}
+
+	for _, endpoint := range endpoints {
+		endpointPort, _ := endpoint.Port()
+		endpointIP := net.ParseIP(endpoint.IP()).To4()
+		ipVal := binary.BigEndian.Uint32(endpointIP)
+		portVal := uint16(endpointPort)
+		group = group.Bucket().Weight(100).
+			LoadReg(int(serviceIPReg), ipVal).
+			LoadRegRange(int(servicePortReg), uint32(portVal), binding.Range{0, 15}).
+			LoadRegRange(int(serviceLearnReg), lbResultMark, binding.Range{16, 31}).
+			ResubmitToTable(resubmitTableID).
```

Comment:
Ditto.

Reply:
I think there is no corresponding implementation in ofnet? I would like to add a TODO here for now.